react/src/components/tooltip/SprkTooltip.js

```diff
@@ -0,0 +1,194 @@
+    this.handleWindowClick = this.handleWindowClick.bind(this);
+    this.setPositioningClass = this.setPositioningClass.bind(this);
+
+    this.triggerRef = React.createRef();
+    this.tooltipRef = React.createRef();
+  }
+
+  componentDidMount() {
+    window.addEventListener('keydown', this.handleWindowKeydown);
+    window.addEventListener('click', this.handleWindowClick);
+
+    this.setPositioningClass();
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener('keydown', this.handleWindowKeydown);
+    window.removeEventListener('click', this.handleWindowClick);
+  }
+
+  setPositioningClass() {
+    let trigger = this.triggerRef.current;
+
+    const elemX = trigger.getBoundingClientRect().left;
+    const elemY = trigger.getBoundingClientRect().top;
+
+    const viewportWidth = window.innerWidth;
+    const viewportHeight = window.innerHeight;
+
+    // 328 is the default max-width
+    let maxWidth = 328;
+    let calculatedWidth = maxWidth;
+
+    if (elemX > viewportWidth / 2) {
+      // the left edge of the button + the width of the button + the border
+      calculatedWidth = elemX + 16 + 16;
+      if (elemY > viewportHeight / 2) {
+        this.setState({ position: 'topleft' })
+      } else {
+        this.setState({ position: 'bottomleft' })
+      }
+    } else {
+      // the width of the viewport less the left edge of the button + the border
+      calculatedWidth = viewportWidth - elemX + 16;
+      if (elemY > viewportHeight / 2) {
+        this.setState({ position: 'topright' })
+      } else {
+        this.setState({ position: 'bottomright' })
+      }
+    }
+
+    if (calculatedWidth < maxWidth){
+      // overwrite the width if there's not enough room to display it
+      this.tooltipRef.current.setAttribute('style', 'width:' + calculatedWidth + "px");
+    }
+  }
+
+  handleWindowKeydown(e) {
+    if (e.key === 'Escape' || e.key === 'Esc' || e.keyCode === 27) {
+      this.setState({
+        isToggled: false
+      })
+    }
+  }
+
+  handleWindowClick(e) {
+    if (this.state.isToggled) {
+      if (!this.tooltipRef.current.contains(e.target)
+        && !this.triggerRef.current.contains(e.target)) {
+        this.setState({
+          isToggled: false
+        });
+      }
+    }
+  }
+
+  toggle() {
+    this.setState(prevState => ({
+      isToggled: !prevState.isToggled
+    }))
+  }
+
+  render() {
+    const {
+      children,
+      idString,
+      triggerIconType,
+      iconAdditionalClasses,
+      additionalClasses,
+      analyticsString,
+      isDefaultOpen,
+      id,
+      ...other
+    } = this.props;
+
+    return (
+      <span {...other} className='sprk-c-Tooltip__container' data-id={idString}>
+        <button
+          ref={this.triggerRef}
+          onClick={this.toggle}
+          onMouseOver={this.setPositioningClass}
+          onFocus={this.setPositioningClass}
+          className={classnames({
+            'sprk-c-Tooltip__trigger': true,
```

Comment:
Don't need to explicitly set this to true. Can be `classNames('sprk-c-Tooltip__trigger', {'toggled stuff'});`